modules/gmake/src/proj.lua

```diff
@@ -0,0 +1,677 @@
+	end
+end
+
+
+---
+-- Prints the CXX flags for the given project.
+--
+-- @param prj
+--  project to print CXX flags for
+---
+function proj.cxxFlags(prj)
+	local toolset = prj.compiler
+	local flags = toolset.getCxxFlags(prj)
+
+	if #flags > 0 then
+		wl('ALL_CXXFLAGS += $(CXXFLAGS) $(ALL_CPPFLAGS) %s', table.concat(flags, ' '))
+	else
+		wl('ALL_CXXFLAGS += $(CXXFLAGS) $(ALL_CPPFLAGS)')
+	end
+end
+
+
+---
+-- Prints the linker flags for the given project.
+--
+-- @param prj
+--  project to print linker flags for
+---
+function proj.linkFlags(prj)
+	local toolset = prj.linker
+	local flags = toolset.getLinkerFlags(prj)
+
+	if #flags > 0 then
+		wl('ALL_LDFLAGS = $(LDFLAGS) %s', table.concat(flags, ' '))
+	else
+		wl('ALL_LDFLAGS = $(LDFLAGS)')
+	end
+end
+
+
+---
+-- Prints all configurations for the project.
+--
+-- @param prj
+--  project to print configurations for
+---
+function proj.configurations(prj)
+	wl('# Configuration-level overrides')
+	for _, cfg in ipairs(prj.configs) do
+		local name = cfg.name:lower()
+
+		local conditional = nil
+		if _ == 1 then
+			conditional = 'ifeq ($(config), ' .. name .. ')'
+		else
+			conditional = 'else ifeq ($(config), ' .. name .. ')'
+		end
+
+		wl(conditional)
+
+		export.indent()
+		premake.callArray(proj.elements.configurations, cfg)
+		export.outdent()
+	end
+
+	if #prj.configs > 0 then
+		wl('endif')
+	end
+
+	wl()
+end
+
+
+---
+-- Prints the target directory for the given configuration.
+--
+-- @param cfg
+--  configuration to print target directory of
+---
+function proj.configTargetDir(cfg)
+	-- TODO: Check if this there is an override of the default dir before writing the TARGETDIR
+	local prj = cfg.project
+	local cfgName = cfg.name
+	local configTargetDir = path.getRelative(prj.location, path.join(prj.workspace.location, 'bin', prj.name, cfgName))
+	wl('TARGETDIR = %s', configTargetDir)
+end
+
+
+---
+-- Prints the intermediates directory for the given configuration.
+--
+-- @param cfg
+--  configuration to print intermediates directory of
+---
+function proj.configIntermediateDir(cfg)
+	-- TODO: Check if this there is an override of the default dir before writing the OBJDIR
+	local prj = cfg.project
+	local cfgName = cfg.name
+	local intDir = path.getRelative(prj.location, path.join(prj.workspace.location, 'obj', prj.name, cfgName))
+	wl('OBJDIR = %s', intDir)
+end
+
+
+---
+-- Prints the target name for the given configuration.
+--
+-- @param cfg
+--  configuration to print target name of
+---
+function proj.configTargetName(cfg)
+	--- TODO: Override target name
+end
+
+
+---
+-- Prints the defines for the given configuration.
+--
+-- @param cfg
+--  configuration to print defines of
+---
+function proj.configDefines(cfg)
+	local defs = cfg.defines
+	if defs ~= nil and #defs > 0 then
+		local defString = table.concat(table.map(defs, function(key, value)
+			return '-D' .. value
+		end), ' ')
+		wl('DEFINES += %s', defString)
+	end
+end
```

Comment:
This could be merged with `project.defines()` above since the logic is the same. Could call the same function, just passing in `prj` to one and `cfg` for the other.

Reply:
Not quite.  Configuration level is +=, whereas project level is =.  Will split the logic up though as previously done.

Reply:
I'm…pretty sure that you can use `+=` even on a new variable? Or, you can check `cfg.project`, which will be set if the object is a project configuration, and `nil` if it is a project.

I want to avoid the situation where there is one implementation for the project, another for project configurations—and another for file-level configurations?—which are basically identical and just getting their data from variables with different names ("prj", "cfg") but all the same fields? So want to make sure it gets off to the right start, with a minimum of duplication,